src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -182,16 +182,29 @@ public void deleteShardDirectorySafe(ShardId shardId, @IndexSettings Settings in
         assert indexSettings != ImmutableSettings.EMPTY;
         final Path[] paths = shardPaths(shardId);
         logger.trace("deleting shard {} directory, paths: [{}]", shardId, paths);
-        try (Closeable lock = shardLock(shardId)) {
-            IOUtils.rm(paths);
-            if (hasCustomDataPath(indexSettings)) {
-                Path customLocation = resolveCustomLocation(indexSettings, shardId);
-                logger.trace("deleting custom shard {} directory [{}]", shardId, customLocation);
-                IOUtils.rm(customLocation);
-            }
-            logger.trace("deleted shard {} directory, paths: [{}]", shardId, paths);
-            assert FileSystemUtils.exists(paths) == false;
+        try (ShardLock lock = shardLock(shardId)) {
+            deleteShardDirectorySafe(lock, indexSettings);
+        }
+    }
+
+    /**
+     * Deletes a shard data directory.
+     *
+     * @param lock the shards lock
+     * @throws IOException if an IOException occurs
+     */
+    public void deleteShardDirectorySafe(ShardLock lock, @IndexSettings Settings indexSettings) throws IOException {
+        assert indexSettings != ImmutableSettings.EMPTY;
```

Comment:
can we check and except if the lock is not acquired? Maybe also rename the method name to deleteShardDirectoryUnderLock, not sure. We should also document the fact that the shard is expected to be locked.